### ISSUE TITLE
Linter crashes on print statement

### ISSUE THREAD

**Issue by gvanrossum** [davidhalter/jedi#662]
I understand the linter is a work in progress. Is it supposed to work at all under Python 2? When running the linter as follows it fails on a file as simple as `print 'hello world'`:

```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/local/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/Users/guido/src/jedi/jedi/__main__.py", line 48, in <module>
    _start_linter()
  File "/Users/guido/src/jedi/jedi/__main__.py", line 31, in _start_linter
    for error in jedi.Script(path=path)._analysis():
  File "jedi/api/__init__.py", line 573, in _analysis
    try_iter_content(self._evaluator.goto_definition(node))
  File "jedi/evaluate/__init__.py", line 423, in goto_definition
    types = self.eval_element(call)
  File "jedi/evaluate/__init__.py", line 257, in eval_element
    return self._eval_element_cached(element)
  File "jedi/evaluate/cache.py", line 41, in wrapper
    rv = function(obj, *args, **kwargs)
  File "jedi/evaluate/__init__.py", line 261, in _eval_element_cached
    return self._eval_element_not_cached(element)
  File "jedi/debug.py", line 53, in wrapper
    return func(*args, **kwargs)
  File "jedi/evaluate/__init__.py", line 309, in _eval_element_not_cached
    types = precedence.calculate_children(self, element.children)
  File "jedi/evaluate/precedence.py", line 47, in calculate_children
    right = next(iterator)
StopIteration
```

I tried this on the linter branch too, no difference.

**Comment by davidhalter** [comment]
@gvanrossum A bit of bad luck. I actually forgot to handle the print statement in the parser in Python 2. It should be working now.

I'd be happy to listen to your feedback as well as a comparison to mypy!

 BTW PEP 484 type hints will be working very soon -> #661

**Comment by gvanrossum** [comment]
Thanks for the quick fix!

Glad to see you're moving forward with PEP 484.

Note that mypy also implements a Python-2-compatible syntax for function
annotations. For example, instead of
```
def add(a: int, b: int) -> int:
    return a + b
```
you would write
```
def add(a, b):
    # type: (int, int) -> int
    return a + b
```
I was actually hoping to be able to use Jedi to infer such signatures from
code that doesn't have any annotations (or not enough) -- do you think that
would be possible with the current Jedi API?

**Comment by davidhalter** [comment]
With the current API it's not possible, but using a bit of the internals, this is possible. Some of the internals are a bit ugly and maybe you don't understand it.

https://gist.github.com/davidhalter/9bde6d0985bd71b103d3

I have tested it on the docopt library itself. Run:

```
 python pep484transform.py path/to/docopt.py
```

Let me know what you think. Glad to help if you're stuck.

**Comment by gvanrossum** [comment]
Wow. For small files this is pretty cool! I ran it on the latest docopt.py
but I get a traceback:

python pep484transform.py ../docopt/docopt.py
Traceback (most recent call last):
  File "pep484transform.py", line 74, in <module>
    print(transform(result['<file>'], result['--debug']))
  File "pep484transform.py", line 35, in transform
    annotation = types_to_string(types, abspath(file_name))
  File "pep484transform.py", line 43, in types_to_string
    transformed = set(type_to_string(typ, path) for typ in types)
  File "pep484transform.py", line 43, in <genexpr>
    transformed = set(type_to_string(typ, path) for typ in types)
  File "pep484transform.py", line 56, in type_to_string
    if str(typ.get_parent_until().name) == 'builtins':
  File "/Users/guido/src/jedi/jedi/_compatibility.py", line 124, in <lambda>
    return lambda *args, **kwargs: self.func(obj, *args, **kwargs)
  File "/Users/guido/src/jedi/jedi/parser/tree.py", line 127, in
get_parent_until
    while scope.parent is not None:
AttributeError: 'ListComprehension' object has no attribute 'parent'

Finally -- I like that Jedi has a round-trippable parser too. I'm a tiny
bit sad though that there's no acknowledgment of its author -- pgen2 is my
work (intended to be round-trippable as I wrote it to be the basis of
2to3), which you copied. It's open source, but would it hurt to add a thank
you to the README? Without it Jedi wouldn't exist. (You technically also
need a copy of the PSF license, but nobody cares. :-)


On Fri, Dec 18, 2015 at 10:00 AM, Dave Halter <notifications@github.com>
wrote:

> With the current API it's not possible, but using a bit of the internals,
> this is possible. Some of the internals are a bit ugly and maybe you don't
> understand it.
>
> https://gist.github.com/davidhalter/9bde6d0985bd71b103d3
>
> I have tested it on the docopt library itself. Run:
>
>  python pep484transform.py path/to/docopt.py
>
> Let me know what you think. Glad to help if you're stuck.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/davidhalter/jedi/issues/662#issuecomment-165857511>.
>



-- 
--Guido van Rossum (python.org/~guido)

**Comment by davidhalter** [comment]
Oh. Sorry I forgot to commit and push the linter branch.

**Comment by gvanrossum** [comment]
NP, that works now. I still occasionally get very long tracebacks -- are
you interesting in seeing those? I presume you can generate them easily
yourself by pointing it at e.g. the Python 3 stdlib or other large code
bases. Have you thought about my concern regarding the origins of pgen2?

On Fri, Dec 18, 2015 at 2:19 PM, Dave Halter <notifications@github.com>
wrote:

> Oh. Sorry I forgot to commit and push the linter branch.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/davidhalter/jedi/issues/662#issuecomment-165910075>.
>



-- 
--Guido van Rossum (python.org/~guido)

**Comment by davidhalter** [comment]
Cool! 

> I still occasionally get very long tracebacks -- are you interesting in seeing those?

Unfortunately there are still quite a few issues. Tracebacks are not needed, I will just ping you again when I'm finished with testing. This will be at least 1-2 months, though (I honestly haven't tested this against anything else than docopt).

> Have you thought about my concern regarding the origins of pgen2?

Not at all, I haven't actually heard them. Could you please enlighten me?

**Comment by gvanrossum** [comment]
I'd love to help you beta-test this when you're further along. It may have
real promise.

The other issue is as follows (sorry, I hid it below a very long traceback
in a previous message):

Finally -- I like that Jedi has a round-trippable parser too. I'm a tiny
bit sad though that there's no acknowledgment of its author -- pgen2 is my
work (intended to be round-trippable as I wrote it to be the basis of
2to3), which you copied. It's open source, but would it hurt to add a thank
you to the README? Without it Jedi wouldn't exist. (You technically also
need a copy of the PSF license, but nobody cares. :-)

On Sat, Dec 19, 2015 at 3:11 AM, Dave Halter <notifications@github.com>
wrote:

> Cool!
>
> I still occasionally get very long tracebacks -- are you interesting in
> seeing those?
>
> Unfortunately there are still quite a few issues. Tracebacks are not
> needed, I will just ping you again when I'm finished with testing. This
> will be at least 1-2 months, though (I honestly haven't tested this against
> anything else than docopt).
>
> Have you thought about my concern regarding the origins of pgen2?
>
> Not at all, I haven't actually heard them. Could you please enlighten me?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/davidhalter/jedi/issues/662#issuecomment-165976998>.
>



-- 
--Guido van Rossum (python.org/~guido)

**Comment by davidhalter** [comment]
>  Finally -- I like that Jedi has a round-trippable parser too. I'm a tiny bit sad though that there's no acknowledgment of its author -- pgen2 is my work (intended to be round-trippable as I wrote it to be the basis of 2to3), which you copied. It's open source, but would it hurt to add a thank you to the README? Without it Jedi wouldn't exist. (You technically also need a copy of the PSF license, but nobody cares. :-)

I actually have thought about this before. Sorry for not doing it back  when 0.9.0 came out. I was a bit confused by all the licensing stuff and I was planning to fix this (and include the PSF license as well), but I have always been side tracked.

One of the issues I had back when I last released was that I was not sure of how to include the PSF license. Do I just have to append it to MIT? Is it now not fully MIT licensed anymore? Would it be possible to dual license it?

Now I'm actually glad that you're concern wasn't like "I wrote it and it sucks" :-) That would have meant a painful rewrite, haha.

**Comment by gvanrossum** [comment]
Only a lawyer can tell you what to do with the license. But I would really
appreciate it if you acknowledged in your README where your parser comes
from, and that I wrote it.

On Sat, Dec 19, 2015 at 8:34 AM, Dave Halter <notifications@github.com>
wrote:

> Finally -- I like that Jedi has a round-trippable parser too. I'm a tiny
> bit sad though that there's no acknowledgment of its author -- pgen2 is my
> work (intended to be round-trippable as I wrote it to be the basis of
> 2to3), which you copied. It's open source, but would it hurt to add a thank
> you to the README? Without it Jedi wouldn't exist. (You technically also
> need a copy of the PSF license, but nobody cares. :-)
>
> I actually have thought about this before. Sorry for not doing it back
> when 0.9.0 came out. I was a bit confused by all the licensing stuff and I
> was planning to fix this (and include the PSF license as well), but I have
> always been side tracked.
>
> One of the issues I had back when I last released was that I was not sure
> of how to include the PSF license. Do I just have to append it to MIT? Is
> it now not fully MIT licensed anymore? Would it be possible to dual license
> it?
>
> Now I'm actually glad that you're concern wasn't like "I wrote it and it
> sucks" :-) That would have meant a painful rewrite, haha.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/davidhalter/jedi/issues/662#issuecomment-166001829>.
>



-- 
--Guido van Rossum (python.org/~guido)

**Comment by davidhalter** [comment]
@gvanrossum I just added an acknowledgement section to the README. Sorry for not mentioning you previously. I wasn't even sure that you wrote the whole thing :) Hope this is what you would have liked in the beginning.

https://github.com/davidhalter/jedi/tree/linter#acknowledgements

**Comment by gvanrossum** [comment]
That's great! Glad to see the fix.

On Fri, Dec 25, 2015 at 7:35 PM, Dave Halter <notifications@github.com>
wrote:

> @gvanrossum <https://github.com/gvanrossum> I just added an
> acknowledgement section to the README. Sorry for not mentioning you
> previously. I wasn't even sure that you wrote the whole thing :) Hope this
> is what you would have liked in the beginning.
>
> https://github.com/davidhalter/jedi/tree/linter#acknowledgements
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/davidhalter/jedi/issues/662#issuecomment-167272548>.
>



-- 
--Guido van Rossum (python.org/~guido)

**Comment by davidhalter** [comment]
With Jedi 0.17.0 being released, I'm going to close this linter issue. The linter was discontinued quite a long time ago. Back when I started working on it, mypy, pyre, pyright and pytype weren't around. Jedi is now almost feature complete and has no open bugs, so I've decided to ditch the linter, it's either not fast enough or not precise enough.

Nowadays with annotations it's pretty easy to type check code. I still think there's value in writing a linter that does not need type annotations, but there's also progress on this in the above tools.

I have realized that there is a fundamental limitations Jedi has: There is no clever caching architecture in place. Of course I could add that, but if I ever do it, I will probably re-implement Jedi in Rust to learn something new and make it faster. (Python is of course still a very awesome language :))
